libraries/AP_HAL_Linux/AnalogIn_Imx_IOMCU.hpp

```diff
@@ -0,0 +1,109 @@
+#pragma once
+
+#include "AP_HAL_Linux.h"
+#include "Imx_IOMCU.hpp"
+
+extern const AP_HAL::HAL &hal;
+
+class AnalogSource_Imx_IOMCU : public AP_HAL::AnalogSource {
+public:
+	friend class AnalogIn_Imx_IOMCU;
+	AnalogSource_Imx_IOMCU(int16_t pin)
+	{
+		_pin = pin;
+		_value = 0;
+	}
+	float read_average()
+	{
+		return read_latest();
+	}
+	float read_latest()
+	{
+		return _value;
+	}
+	void set_pin(uint8_t p)
+	{
+		_pin = p;
+	}
+	void set_stop_pin(uint8_t p) {}
+	void set_settle_time(uint16_t settle_time_ms) {}
+	float voltage_average()
+	{
+		return _value;
+	}
+	float voltage_latest()
+	{
+		return _value;
+	}
+	float voltage_average_ratiometric()
+	{
+		return _value;
+	}
+private:
+	int16_t _pin;
+	float _value;
+};
+
+class AnalogIn_Imx_IOMCU : public AP_HAL::AnalogIn {
+public:
+	AnalogIn_Imx_IOMCU(Imx_IOMCU *m)
+	{
+		mcu = m;
+	}
+
+	void init() override
+	{
+		hal.scheduler->register_timer_process(FUNCTOR_BIND_MEMBER(&AnalogIn_Imx_IOMCU::_update, void));
+	}
+	AP_HAL::AnalogSource *channel(int16_t n) override
+	{
+		for (uint8_t j = 0; j < 2; j++) {
+			if (_channels[j] == nullptr) {
+				_channels[j] = new AnalogSource_Imx_IOMCU(n);
+				return _channels[j];
+			}
+		}
+
+		hal.console->printf("Out of analog channels\n");
+		return nullptr;
+	}
+
+	/* Board voltage is not available */
+	float board_voltage() override { return 5.0f; }
+
+private:
+	void _update()
+	{
```

Comment:
this method is called from a different thread than the others. You need to synchronize access to the shared data. See how the other implementations use `WITH_SEMAPHORE()` to protect the data.

Reply:
here,i think standard variable assignment,is lock free,i do not need any SEMAPHORE or mutex.
is it right?

Reply:
It may or may not be depending on the CPU. And it may require a barrier for other cores to see the updated value (one reason why we have atomic types in stdatomic.h and c++'s atomic).

Reply:
ok,thanks